### PR TITLE
add Sparks (SPK) support.

### PR DESCRIPTION
**Pull request by zawasp**
Added NeoScrypt code from https://github.com/mogwaicoin/NeoScryptCSharp
It needs 'unsafe' code flag.

### CODE REVIEW

File: NBitcoin.TestFramework/NBitcoin.TestFramework.csproj [hunk 1]
```diff
@@ -21,7 +21,8 @@
 		<GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
 		<GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
     <NoWarn>1591;1573;1572;1584;1570;3021</NoWarn>
-	</PropertyGroup>
+    <!--<AllowUnsafeBlocks>true</AllowUnsafeBlocks>-->
```
Comment (NicolasDorier) [inline NBitcoin.TestFramework/NBitcoin.TestFramework.csproj]: remove those changes

**Comment by zawasp** [comment]
> Also, mogwai is already supported, don't the library have that already?

It is implemented, but their CalculateHash method doesn't actually calculate it (code is commented out). The NeoScrypt hashing algo needs to be added to support Mogwai, GBX, SPK (and maybe others in the future).

**Comment by zawasp** [comment]
> Please remove code using pointers.
> 
> I don't have time to review the alg implementation, but if there is a bug inside and pointers are used, it can open to bad attack vector.

I know that, however it's the only C# NeoScrypt library I found.

File: NBitcoin/NBitcoin.csproj [hunk 2]
```diff
@@ -24,7 +24,8 @@
 		<TargetFrameworks Condition="'$(BuildingInsideVisualStudio)' == 'true'">netstandard2.0</TargetFrameworks>
 		<NoWarn>1591;1573;1572;1584;1570;3021</NoWarn>
 		<GenerateDocumentationFile>true</GenerateDocumentationFile>
-	</PropertyGroup>
+    <!--<AllowUnsafeBlocks>true</AllowUnsafeBlocks>-->
```
[the remark this answers is not in the document]
Comment (zawasp) [reply NBitcoin/NBitcoin.csproj]: They were removed in the last commits

File: NBitcoin.Tests/NBitcoin.Tests.csproj [hunk 3]
```diff
@@ -10,14 +10,14 @@
 	  <TargetFrameworks Condition="'$(BuildingInsideVisualStudio)' == 'true'">netcoreapp2.1</TargetFrameworks>
 	  <!--<TargetFrameworks Condition="'$(BuildingInsideVisualStudio)' == 'true' And '$(MSBuildRuntimeType)' == 'Core' ">netcoreapp2.1</TargetFrameworks>-->
 	  <!--<NoWarn>CS0618</NoWarn>-->
+    <!--<AllowUnsafeBlocks>true</AllowUnsafeBlocks>-->
```
[the remark this answers is not in the document]
Comment (zawasp) [reply NBitcoin.Tests/NBitcoin.Tests.csproj]: Removed in the last commits

**Comment by NicolasDorier** [comment]
Do you need `NeoScrypt` ? Because afaik mogwai is already here and also depends on it in theory.

**Comment by zawasp** [comment]
> Do you need `NeoScrypt` ? Because afaik mogwai is already here and also depends on it in theory.

It's the hashing algo used by both mogwai, GBX and SPK. Without it I'm not sure GetHash will work.

**Comment by NicolasDorier** [comment]
I don't try altcoins code, maybe one day I will do it on a VM and remove those which are not fully implemented. No unsafe code because this is a security risk.

**Comment by zawasp** [comment]
> I don't try altcoins code, maybe one day I will do it on a VM and remove those which are not fully implemented. No unsafe code because this is a security risk.

Alright, I will try to rewrite the NeoScrypt lib or find another one which doesn't require unsafe code. I'll close this one for now.

**Pull request closed without merging**